global.go

```diff
@@ -0,0 +1,30 @@
+package netdicom
+
+var allTransferSyntaxes = []string{
```

Comment:
Could you add a comment on how this list is composed? I mean, how do we determine if a transfer syntax is supported or not by our software?

Reply:
I took from https://www.dicomlibrary.com/dicom/transfer-syntax/, they are just sorted in alphabetical order

Reply:
So they are not the list of supported transfer syntax, so there is no guarantee that the picked transfer syntax is supported by our software at all?

In this case, I think we should change the pick transfer syntax to using a black list to reject the known not supported transfer syntax instead of maintaining this list. The logic should looks like:


```go
func pickFirstPreferedTransferSyntax(proposedTransferSyntaxUIDs []string) string {
	var blacklist = map[string]struct{}{
		"1.2.840.10008.1.2.4.90": struct{}{}, // JPEG 2000 Lossless
		// all know known-to-not-supported transfer syntaxes
	}
	for _, transferSyntaxUID := range proposedTransferSyntaxUIDs {
		if _, ok := blacklist[transferSyntaxUID]; ok {
			continue
		}
		return transferSyntaxUID
	}
}
```

Could you see if it make more sense?

Reply:
it's the same. If the proposed transfer syntaxes are all not supported, we should use the default transfer syntax 
```
"1.2.840.10008.1.2",       // Implicit VR Little Endian
```
which shall be supported by every conformant DICOM Implementation.

Reply:
Even if we doesn't support JPEG2000, we should let them send those images to us. I tested with pynetdicom and dcmtk, both of them used the same approached.

Reply:
> it's the same

The result should be the same, the different is we don't have to maintain the `allTransferSyntaxes` list, but only the list of known-not-supported transfer syntax. It also helps with code readability as we know exactly which transfer syntaxes are not supported.

> Even if we doesn't support JPEG2000, we should let them send those images to us

That's correct.  In the example codes above, the `continue` statement will make `pickFirstPreferedTransferSyntax` picked the next proposed transfer syntax.

Reply:
what happens if they propose only one transfer syntax, but we don't support it? e.g JPEG2000

Reply:
> We still need to let them send these images, right?

Yes, as you said above

> If the proposed transfer syntaxes are all not supported, we should use the default transfer syntax

The above example code is just for illustration. The idea is use a blacklist with clearly defined not-supported transfer syntaxes blacklist instead of using an unverified whitelist.

Reply:
I prefer this approach because it's general-propose and we can prefer the transfer syntax in order instead of just pick the first proposed transfer syntax they offer us.

Reply:
> we can prefer the transfer syntax in order instead of just pick the first proposed transfer syntax they offer us.

From what I understand from the logic of `pickFirstPreferedTransferSyntax` the picked result should be the same with white list or blacklist. For example:

* `allTransferSyntaxes` is `[A, B, C]`
* `unsupportedTransferSyntax` is `[D]`
* `proposedTransferSyntaxes` is `[D, B]`

with the current codes

```
for _, syntaxUID := range `[A, B, C` {
 		for _, proposed := range `[D, B]` {
 			if syntaxUID == proposed {
 				return syntaxUID
 			}
 		}
 	}
```

the picked transfer syntax will be `B`.

if we used blacklist

```
for _, syntaxID := range `[D, B]` {
    if _, ok := blacklist[syntaxID]; ok {
         continue
   }
   return syntaxID
}
```

The picked transfer syntax will still be `B`, the end result is no different. 

Please review and give a specific example when the whitelist approach gives a different/better result.

Reply:
I realized after posting the previous comment with the current logic

```
for _, syntaxUID := range `[A, B, C]` {
 		for _, proposed := range `[D, B]` {
 			if syntaxUID == proposed {
 				return syntaxUID
 			}
 		}
 	}
```

will always pick `Implicit VR Little Endian` regardless of the client proposed transfer syntax, as it is supported by most clients. I don't think we should do that, the client proposed the preferred transfer syntax ids for a reason.

Reply:
```
allTransferSyntaxes is [A, B, C]
proposedTransferSyntaxes is [C, A]
```
in this case, we should pick A instead of C because its priority is higher than C.

> will always pick `Implicit VR Little Endian` regardless of the client proposed transfer syntax, as it is supported by most clients. I don't think we should do that, the client proposed the preferred transfer syntax ids for a reason.

Actually, we pick B. `Implicit VR Little Endian` is only used as the last resource when there's no matching syntax between client and server.

Reply:
```
storescp --debug 7250 
D: $dcmtk: storescp v3.6.6 2021-01-14 $
D:
D: DcmDataDictionary: Loading file: /usr/local/Cellar/dcmtk/3.6.6/share/dcmtk/dicom.dic
D: Association Received: localhost
D: setting network send timeout to 60 seconds
D: setting network receive timeout to 60 seconds
D: PDU Type: Associate Request, PDU Length: 430 + 6 bytes PDU header
D:   01  00  00  00  01  ae  00  01  00  00  73  65  67  6d  65  64
D:   5f  61  65  20  20  20  20  20  20  20  6e  6f  76  61  70  61
D:   63  73  20  20  20  20  20  20  20  20  00  00  00  00  00  00
D:   00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
D:   00  00  00  00  00  00  00  00  00  00  10  00  00  15  31  2e
D:   32  2e  38  34  30  2e  31  30  30  30  38  2e  33  2e  31  2e
D:   31  2e  31  20  00  00  67  01  00  00  00  30  00  00  19  31
D:   2e  32  2e  38  34  30  2e  31  30  30  30  38  2e  35  2e  31
D:   2e  34  2e  31  2e  31  2e  31  40  00  00  16  31  2e  32  2e
D:   38  34  30  2e  31  30  30  30  38  2e  31  2e  32  2e  34  2e
D:   39  31  40  00  00  13  31  2e  32  2e  38  34  30  2e  31  30
D:   30  30  38  2e  31  2e  32  2e  31  40  00  00  11  31  2e  32
D:   2e  38  34  30  2e  31  30  30  30  38  2e  31  2e  32  20  00
D:   00  67  03  00  00  00  30  00  00  19  31  2e  32  2e  38  34
D:   30  2e  31  30  30  30  38  2e  35  2e  31  2e  34  2e  31  2e
D:   31  2e  37  40  00  00  16  31  2e  32  2e  38  34  30  2e  31
D:   30  30  30  38  2e  31  2e  32  2e  34  2e  39  31  40  00  00
D:   13  31  2e  32  2e  38  34  30  2e  31  30  30  30  38  2e  31
D:   2e  32  2e  31  40  00  00  11  31  2e  32  2e  38  34  30  2e
D:   31  30  30  30  38  2e  31  2e  32  50  00  00  77  51  00  00
D:   04  00  00  00  00  52  00  00  12  31  2e  32  2e  38  34  30
D:   2e  31  31  34  30  35  31  2e  36  2e  30  53  00  00  04  00
D:   01  00  01  54  00  00  1d  00  19  31  2e  32  2e  38  34  30
D:   2e  31  30  30  30  38  2e  35  2e  31  2e  34  2e  31  2e  31
D:   2e  31  01  00  54  00  00  1d  00  19  31  2e  32  2e  38  34
D:   30  2e  31  30  30  30  38  2e  35  2e  31  2e  34  2e  31  2e
D:   31  2e  37  01  00  55  00  00  0b  4e  6f  76  61  52  61  64
D:   20  38  2e  30
D: Parsing an A-ASSOCIATE PDU
I: Association Received
D: Parameters:
D: ====================== BEGIN A-ASSOCIATE-RQ =====================
D: Our Implementation Class UID:      1.2.276.0.7230010.3.0.3.6.6
D: Our Implementation Version Name:   OFFIS_DCMTK_366
D: Their Implementation Class UID:    1.2.840.114051.6.0
D: Their Implementation Version Name: NovaRad 8.0
D: Application Context Name:    1.2.840.10008.3.1.1.1
D: Calling Application Name:    novapacs
D: Called Application Name:     segmed_ae
D: Responding Application Name:
D: Our Max PDU Receive Size:    16384
D: Their Max PDU Receive Size:  0
D: Presentation Contexts:
D:   Context ID:        1 (Proposed)
D:     Abstract Syntax: =ComputedRadiographyImageStorage
D:     Proposed SCP/SCU Role: SCU
D:     Proposed Transfer Syntax(es):
D:       =JPEG2000
D:       =LittleEndianExplicit
D:       =LittleEndianImplicit
D:   Context ID:        3 (Proposed)
D:     Abstract Syntax: =SecondaryCaptureImageStorage
D:     Proposed SCP/SCU Role: SCU
D:     Proposed Transfer Syntax(es):
D:       =JPEG2000
D:       =LittleEndianExplicit
D:       =LittleEndianImplicit
D: Requested Extended Negotiation: none
D: Accepted Extended Negotiation:  none
D: Requested User Identity Negotiation: none
D: User Identity Negotiation Response:  none
D: ======================= END A-ASSOCIATE-RQ ======================
D: Constructing Associate AC PDU
I: Association Acknowledged (Max Send PDV: 131060)
D: ====================== BEGIN A-ASSOCIATE-AC =====================
D: Our Implementation Class UID:      1.2.276.0.7230010.3.0.3.6.6
D: Our Implementation Version Name:   OFFIS_DCMTK_366
D: Their Implementation Class UID:    1.2.840.114051.6.0
D: Their Implementation Version Name: NovaRad 8.0
D: Application Context Name:    1.2.840.10008.3.1.1.1
D: Calling Application Name:    novapacs
D: Called Application Name:     segmed_ae
D: Responding Application Name: segmed_ae
D: Our Max PDU Receive Size:    16384
D: Their Max PDU Receive Size:  0
D: Presentation Contexts:
D:   Context ID:        1 (Accepted)
D:     Abstract Syntax: =ComputedRadiographyImageStorage
D:     Proposed SCP/SCU Role: SCU
D:     Accepted SCP/SCU Role: Default
D:     Accepted Transfer Syntax: =LittleEndianExplicit
D:   Context ID:        3 (Accepted)
D:     Abstract Syntax: =SecondaryCaptureImageStorage
D:     Proposed SCP/SCU Role: SCU
D:     Accepted SCP/SCU Role: Default
D:     Accepted Transfer Syntax: =LittleEndianExplicit
D: Requested Extended Negotiation: none
D: Accepted Extended Negotiation:  none
D: Requested User Identity Negotiation: none
D: User Identity Negotiation Response:  none
D: ======================= END A-ASSOCIATE-AC ======================
D: DcmDataset::read() TransferSyntax="Little Endian Implicit"
I: Received Store Request
D: ===================== INCOMING DIMSE MESSAGE ====================
D: Message Type                  : C-STORE RQ
D: Presentation Context ID       : 1
D: Message ID                    : 46674
D: Affected SOP Class UID        : ComputedRadiographyImageStorage
D: Affected SOP Instance UID     : 1.2.392.200036.9125.9.0.353320893.1828461056.3248501221
D: Data Set                      : present
D: Priority                      : medium
D: ======================= END DIMSE MESSAGE =======================
D: DcmDataset::read() TransferSyntax="Little Endian Explicit"
```

Reply:
After discussed, we feel like using default `Little Endian Explicit` is a safer bet at this moment.